File: junit/rails/game/state/BooleanStateTest.java
package rails.game.state;

import static org.fest.assertions.api.Assertions.assertThat;
import static org.junit.Assert.*;

import org.junit.Before;
import org.junit.Test;

public class BooleanStateTest {

    private final static String DEFAULT_ID = "Default";
    private final static String INIT_ID = "Init"; 

    private Root root;
    
    private BooleanState stateDefault;
    private BooleanState stateInit;

    @Before
    public void setUp() {
        root = StateTestUtils.setUpRoot();
        stateDefault = BooleanState.create(root,  DEFAULT_ID);
        stateInit = BooleanState.create(root, INIT_ID, true);
        
    }

    @Test
    public void testValue() {
        assertFalse(stateDefault.value());
        assertTrue(stateInit.value());
    }
    
    @Test
    public void testSet() {
        stateDefault.set(true);
        assertTrue(stateDefault.value());
        stateInit.set(false);
        assertFalse(stateInit.value());
    }

    @Test
    public void testSetSameIgnored() {
        stateDefault.set(false);
        stateInit.set(true);
        StateTestUtils.close(root);
        assertThat(StateTestUtils.getLastClosedChangeSet(root).getStates()).doesNotContain(stateDefault, stateInit);
    }
    
    @Test
    public void testUndoRedo() {
        assertFalse(stateDefault.value());
        stateDefault.set(true);
        assertTrue(stateDefault.value());
        StateTestUtils.close(root);
        assertThat(StateTestUtils.getLastClosedChangeSet(root).getStates()).contains(stateDefault);
        StateTestUtils.undo(root);
        assertFalse(stateDefault.value());
        StateTestUtils.redo(root);
        assertTrue(stateDefault.value());
    }

}

File: dotnet/GameStateTest/BooleanStateTest.cs
using GameLib.Net.Game.State;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace GameStateTest
{
    [TestClass]
    public class BooleanStateTest
    {
        private const string DEFAULT_ID = "Default";
        private const string INIT_ID = "Init";

        private Root root;

        private BooleanState stateDefault;
        private BooleanState stateInit;

        [TestInitialize]
        public void SetUp()
        {
            root = StateTestUtils.SetUpRoot();
            stateDefault = BooleanState.Create(root, DEFAULT_ID);
            stateInit = BooleanState.Create(root, INIT_ID, true);

        }

        [TestMethod]
        public void TestValue()
        {
            Assert.IsFalse(stateDefault.Value);
            Assert.IsTrue(stateInit.Value);
        }

        [TestMethod]
        public void TestSet()
        {
            stateDefault.Set(true);
            Assert.IsTrue(stateDefault.Value);
            stateInit.Set(false);
            Assert.IsFalse(stateInit.Value);
        }

        [TestMethod]
        public void TestSetSameIgnored()
        {
            stateDefault.Set(false);
            stateInit.Set(true);
            StateTestUtils.Close(root);
            var a = StateTestUtils.GetPreviousChangeSet(root).GetStates();
            Assert.IsTrue(!(a.Contains(stateDefault) || a.Contains(stateInit)));
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).doesNotContain(stateDefault, stateInit);
        }

        [TestMethod]
        public void TestUndoRedo()
        {
            Assert.IsFalse(stateDefault.Value);
            stateDefault.Set(true);
            Assert.IsTrue(stateDefault.Value);
            StateTestUtils.Close(root);
            Assert.IsTrue(StateTestUtils.GetPreviousChangeSet(root).GetStates().Contains(stateDefault));
            //assertThat(StateTestUtils.getPreviousChangeSet(root).getStates()).contains(stateDefault);
            StateTestUtils.Undo(root);
            Assert.IsFalse(stateDefault.Value);
            StateTestUtils.Redo(root);
            Assert.IsTrue(stateDefault.Value);
        }
    }
}
